File: v8/suitecrm_client.php
#!/usr/bin/env php
<?php
/**
 * SuiteCRM / SinergiaCRM API V8 Client (PHP CLI)
 * Connects using client_credentials grant, gets an access token, and lists contacts.
 *
 * Usage: php suitecrm_client.php [--all]
 *
 * Configuration is read from .env in the same directory.
 * Copy .env.example to .env and set your credentials.
 */

define('APP_DIR', __DIR__);
define('JSONAPI_MIME', 'application/vnd.api+json');

function loadEnv(): array
{
    $envFile = APP_DIR . '/.env';
    $vars = [];
    if (file_exists($envFile)) {
        foreach (file($envFile, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES) as $line) {
            $line = trim($line);
            if ($line === '' || $line[0] === '#') {
                continue;
            }
            [$key, $value] = explode('=', $line, 2);
            $vars[trim($key)] = trim($value);
        }
    }
    return $vars;
}

$env = loadEnv();
if (empty($env)) {
    fwrite(STDERR, "Error: .env file not found or empty. Copy .env.example to .env and configure it.\n");
    exit(1);
}

define('SUITECRM_BASE_URL', rtrim($env['SUITECRM_BASE_URL'] ?? 'http://localhost:8000', '/'));
define('OAUTH2_CLIENT_ID', $env['OAUTH2_CLIENT_ID'] ?? '');
define('OAUTH2_CLIENT_SECRET', $env['OAUTH2_CLIENT_SECRET'] ?? '');

function getAccessToken(): string
{
    $ch = curl_init(SUITECRM_BASE_URL . '/Api/access_token');
    curl_setopt_array($ch, [
        CURLOPT_POST => true,
        CURLOPT_POSTFIELDS => json_encode([
            'grant_type' => 'client_credentials',
            'client_id' => OAUTH2_CLIENT_ID,
            'client_secret' => OAUTH2_CLIENT_SECRET,
        ]),
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_HTTPHEADER => ['Content-Type: ' . JSONAPI_MIME, 'Accept: ' . JSONAPI_MIME],
    ]);
    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    curl_close($ch);

    if ($httpCode !== 200) {
        throw new RuntimeException("Auth failed ($httpCode): $response");
    }
    return json_decode($response, true)['access_token'];
}

function apiGet(string $endpoint, string $token, array $params = []): array
{
    $url = SUITECRM_BASE_URL . $endpoint;
    if ($params) {
        $url .= '?' . http_build_query($params);
    }
    $ch = curl_init($url);
    curl_setopt_array($ch, [
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_HTTPHEADER => [
            'Content-Type: ' . JSONAPI_MIME,
            'Accept: ' . JSONAPI_MIME,
            'Authorization: Bearer ' . $token,
        ],
    ]);
    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    curl_close($ch);

    if ($httpCode !== 200) {
        throw new RuntimeException("API error ($httpCode): $response");
    }
    return json_decode($response, true);
}

function fetchContacts(string $token, int $pageSize = 20, ?int $maxPages = null): array
{
    $contacts = [];
    $pageNumber = 1;

    while (true) {
        $data = apiGet('/Api/V8/module/Contacts', $token, [
            'page[number]' => $pageNumber,
            'page[size]' => $pageSize,
        ]);
        $pageContacts = $data['data'] ?? [];
        if (empty($pageContacts)) {
            break;
        }
        $contacts = array_merge($contacts, $pageContacts);
        $totalPages = $data['meta']['total-pages'] ?? 1;

        echo sprintf("  Page %d/%d (%d contacts)\n", $pageNumber, $totalPages, count($pageContacts));

        if ($pageNumber >= $totalPages || ($maxPages !== null && $pageNumber >= $maxPages)) {
            break;
        }
        $pageNumber++;
    }

    return [$contacts, $totalPages];
}

echo "=== SuiteCRM API V8 Client ===\n";
echo "Base URL: " . SUITECRM_BASE_URL . "\n\n";

echo "Step 1: Getting access token...\n";
$token = getAccessToken();
echo "  Token: " . substr($token, 0, 50) . "...\n\n";

$allFlag = in_array('--all', $argv ?? []);
$maxPages = $allFlag ? null : 3;
$pageSize = 20;

$label = $allFlag ? 'all pages' : "up to $maxPages pages";
echo "Step 2: Fetching contacts ($label, $pageSize per page)...\n";
[$contacts, $totalPages] = fetchContacts($token, $pageSize, $maxPages);
echo "\n";

$estimatedTotal = $totalPages * $pageSize;
echo sprintf("Contacts fetched: %d of ~%d total\n\n", count($contacts), $estimatedTotal);

$i = 1;
foreach ($contacts as $contact) {
    $attrs = $contact['attributes'] ?? [];
    $contactId = $contact['id'] ?? 'N/A';
    $firstName = $attrs['first_name'] ?? '';
    $lastName = $attrs['last_name'] ?? '';
    $email = $attrs['email1'] ?? '';
    $fullName = $attrs['name'] ?? trim("$firstName $lastName");

    echo sprintf("  %d. ID: %s\n", $i++, $contactId);
    echo sprintf("     Name: %s\n", $fullName);
    if ($email) {
        echo sprintf("     Email: %s\n", $email);
    }
    echo "\n";
}

if (!$allFlag && $estimatedTotal > count($contacts)) {
    echo sprintf("Showing %d of ~%d contacts.\nUse --all to fetch all pages.\n", count($contacts), $estimatedTotal);
}

File: v8/suitecrm_client.py
#!/usr/bin/env python3
"""
SuiteCRM / SinergiaCRM API V8 Client (Python CLI)
Connects using client_credentials grant, gets an access token, and lists contacts.

Configuration is read from .env in the same directory.
Copy .env.example to .env and set your credentials.
"""

import json
import os
import sys
import urllib.request
import urllib.error

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))


def load_env():
    env_file = os.path.join(SCRIPT_DIR, ".env")
    if not os.path.exists(env_file):
        sys.exit("Error: .env file not found. Copy .env.example to .env and configure it.")

    env = {}
    with open(env_file, "r") as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith("#"):
                continue
            key, _, value = line.partition("=")
            env[key.strip()] = value.strip()
    return env


ENV = load_env()
BASE_URL = ENV.get("SUITECRM_BASE_URL", "http://localhost:8000").rstrip("/")
CLIENT_ID = ENV.get("OAUTH2_CLIENT_ID", "")
CLIENT_SECRET = ENV.get("OAUTH2_CLIENT_SECRET", "")
HEADERS_JSONAPI = {
    "Content-Type": "application/vnd.api+json",
    "Accept": "application/vnd.api+json",
}


def get_access_token():
    url = f"{BASE_URL}/Api/access_token"
    body = json.dumps({
        "grant_type": "client_credentials",
        "client_id": CLIENT_ID,
        "client_secret": CLIENT_SECRET,
    }).encode("utf-8")

    req = urllib.request.Request(url, data=body, headers=HEADERS_JSONAPI, method="POST")
    try:
        with urllib.request.urlopen(req) as resp:
            data = json.loads(resp.read().decode("utf-8"))
            return data["access_token"]
    except urllib.error.HTTPError as e:
        error_body = e.read().decode("utf-8")
        raise RuntimeError(f"Auth failed ({e.code}): {error_body}")


def api_get(endpoint, token, params=None):
    if params:
        parts = [f"{urllib.request.quote(str(k))}={urllib.request.quote(str(v))}" for k, v in params.items()]
        url = f"{BASE_URL}{endpoint}?{'&'.join(parts)}"
    else:
        url = f"{BASE_URL}{endpoint}"

    headers = {**HEADERS_JSONAPI, "Authorization": f"Bearer {token}"}
    req = urllib.request.Request(url, headers=headers, method="GET")

    try:
        with urllib.request.urlopen(req) as resp:
            return json.loads(resp.read().decode("utf-8"))
    except urllib.error.HTTPError as e:
        error_body = e.read().decode("utf-8")
        raise RuntimeError(f"API error ({e.code}): {error_body}")


def fetch_contacts(token, page_size=20, max_pages=None):
    all_contacts = []
    page_number = 1

    while True:
        params = {"page[number]": page_number, "page[size]": page_size}
        data = api_get("/Api/V8/module/Contacts", token, params)
        contacts = data.get("data", [])
        if not contacts:
            break

        all_contacts.extend(contacts)
        total_pages = data.get("meta", {}).get("total-pages", 1)
        print(f"  Page {page_number}/{total_pages} ({len(contacts)} contacts)")

        if page_number >= total_pages or (max_pages and page_number >= max_pages):
            break
        page_number += 1

    return all_contacts, total_pages


def main():
    all_flag = "--all" in sys.argv
    max_pages = None if all_flag else 3
    page_size = 20

    print("=== SuiteCRM API V8 Client ===")
    print(f"Base URL: {BASE_URL}")
    print()

    print("Step 1: Getting access token...")
    token = get_access_token()
    print(f"  Token: {token[:50]}...")
    print()

    label = "all pages" if all_flag else f"up to {max_pages} pages"
    print(f"Step 2: Fetching contacts ({label}, {page_size} per page)...")
    contacts, total_pages = fetch_contacts(token, page_size=page_size, max_pages=max_pages)
    print()

    estimated_total = total_pages * page_size
    print(f"Contacts fetched: {len(contacts)} of ~{estimated_total} total")
    print()

    for i, contact in enumerate(contacts, 1):
        attrs = contact.get("attributes", {})
        contact_id = contact.get("id", "N/A")
        first_name = attrs.get("first_name", "")
        last_name = attrs.get("last_name", "")
        email = attrs.get("email1", "")
        full_name = attrs.get("name", f"{first_name} {last_name}".strip())
        print(f"  {i}. ID: {contact_id}")
        print(f"     Name: {full_name}")
        if email:
            print(f"     Email: {email}")
        print()

    if not all_flag and estimated_total > len(contacts):
        print(f"Showing {len(contacts)} of ~{estimated_total} contacts.")
        print("Use --all to fetch all pages.")


if __name__ == "__main__":
    main()
